File: ios/Mailrs/Accounts/MessageBody.swift
import Foundation

/// The readable part of a message, out of its raw bytes.
///
/// Not a full MIME implementation and not trying to be: what a reader
/// needs is the one part worth showing and the text of it. Attachments,
/// signatures and the rest of the tree are left alone.
///
/// Bytes rather than a `String` throughout, and that is the whole
/// reason this exists separately from `MessageHeaders`. A message says
/// what its charset is *inside itself*; decoding it as UTF-8 on the way
/// in — which is what reading a socket into a `String` does — turns
/// every Shift_JIS and windows-1252 message into replacement
/// characters before anything has read the header that would have said
/// so.
enum MessageBody {
    struct Display: Equatable {
        var text: String
        /// Whether `text` is HTML. The caller renders accordingly; a
        /// reader shown raw markup is a reader shown a defect.
        var isHTML: Bool

        static let empty = Display(text: "", isHTML: false)
    }

    /// The part worth showing.
    static func extract(_ raw: Data) -> Display {
        let (headerBytes, body) = split(raw)
        let header = String(decoding: headerBytes, as: UTF8.self)
        return part(header: header, body: body)
    }

    // MARK: - structure

    private static func part(header: String, body: Data) -> Display {
        let type = contentType(header)
        if type.type == "multipart" {
            guard let boundary = type.params["boundary"], !boundary.isEmpty else {
                // A multipart with no boundary cannot be taken apart.
                // Showing the raw source beats showing nothing: the
                // text is usually in there, sandwiched between header
                // blocks a person can read past. Deliberately not via
                // `decoded`, which would throw it away for not being
                // `text/*` — it is not text/* by declaration only.
                return Display(
                    text: string(
                        decodeTransfer(body, as: encoding(header)),
                        charset: type.params["charset"]),
                    isHTML: false)
            }
            return choose(among: pieces(of: body, boundary: boundary), kind: type.subtype)
        }
        return decoded(header: header, body: body)
    }

    /// Which of a multipart's pieces to show.
    ///
    /// `alternative` is the same message written twice, so the choice
    /// is a preference: plain text first, markup only when there is no
    /// plain text. Every other kind — `mixed`, `related`, `signed` — is
    /// a message plus its attachments, and the first piece with
    /// anything readable in it is the message.
    private static func choose(among parts: [Data], kind: String) -> Display {
        let shown = parts.map { piece -> Display in
            let (h, b) = split(piece)
            return part(header: String(decoding: h, as: UTF8.self), body: b)
        }
        if kind == "alternative" {
            if let plain = shown.first(where: { !$0.isHTML && !$0.text.isEmpty }) { return plain }
        }
        return shown.first(where: { !$0.text.isEmpty }) ?? .empty
    }

    /// The pieces between the boundary delimiters.
    ///
    /// Everything before the first delimiter is the preamble and
    /// everything after the closing one is the epilogue; both are
    /// there for mail readers that cannot do MIME at all, and neither
    /// is part of the message.
    static func pieces(of body: Data, boundary: String) -> [Data] {
        let delimiter = Data("--\(boundary)".utf8)
        var out: [Data] = []
        var starts = ranges(of: delimiter, in: body)
        guard !starts.isEmpty else { return [] }
        // The delimiter must begin a line, or a boundary string that
        // happens to appear inside a part cuts it in half.
        //
        // `startIndex`, never 0: every `Data` here is a **slice** of
        // the message, whose indices run from where the slice began.
        // Comparing against 0 reads before the slice, and indexing by
        // `count` reads past its end — which is what crashed four of
        // these tests outright rather than failing them.
        starts = starts.filter { $0 == body.startIndex || body[$0 - 1] == 0x0A }
        for (i, start) in starts.enumerated() {
            let afterDelimiter = start + delimiter.count
            if afterDelimiter + 1 < body.endIndex,
                body[afterDelimiter] == 0x2D, body[afterDelimiter + 1] == 0x2D
            { break }  // the closing delimiter: nothing after it is ours
            guard i + 1 < starts.count else { break }
            let from = skipLine(body, afterDelimiter)
            let to = starts[i + 1]
            if from < to { out.append(body[from..<to]) }
        }
        return out
    }

    // MARK: - leaves

    private static func decoded(header: String, body: Data) -> Display {
        let type = contentType(header)
        // Anything that is not text is not something to show as text.
        // An attached PDF decoded as if it were a message reads as a
        // screen of noise.
        if !type.type.isEmpty && type.type != "text" { return .empty }
        let bytes = decodeTransfer(body, as: encoding(header))
        let text = string(bytes, charset: type.params["charset"])
        return Display(text: text, isHTML: type.subtype == "html")
    }

    static func decodeTransfer(_ body: Data, as cte: String) -> Data {
        switch cte {
        case "base64": return Base64Body.decode(body)
        case "quoted-printable": return QuotedPrintable.decode(body)
        default: return body
        }
    }

    /// Bytes to text, in the charset the message declared.
    ///
    /// UTF-8 when nothing was declared — it is what most mail is, and
    /// it is self-checking, so a wrong guess fails loudly enough to
    /// fall back rather than producing plausible nonsense.
    private static func string(_ bytes: Data, charset: String?) -> String {
        guard let charset, !charset.isEmpty else {
            return String(decoding: bytes, as: UTF8.self)
        }
        let cf = CFStringConvertIANACharSetNameToEncoding(charset as CFString)
        if cf != kCFStringEncodingInvalidId {
            let ns = CFStringConvertEncodingToNSStringEncoding(cf)
            if let text = String(data: bytes, encoding: String.Encoding(rawValue: ns)) {
                return text
            }
        }
        return String(decoding: bytes, as: UTF8.self)
    }

    // MARK: - headers of a part

    struct ContentType: Equatable {
        var type = ""
        var subtype = ""
        var params: [String: String] = [:]
    }

    static func contentType(_ header: String) -> ContentType {
        guard let raw = value(of: "content-type", in: header) else {
            return ContentType(type: "text", subtype: "plain")
        }
        var out = ContentType()
        let fields = splitOnSemicolons(raw)
        let full = fields.first ?? ""
        let halves = full.split(separator: "/", maxSplits: 1)
        out.type = halves.first.map { String($0).lowercased() } ?? ""
        if halves.count > 1 { out.subtype = String(halves[1]).lowercased() }
        for field in fields.dropFirst() {
            guard let eq = field.firstIndex(of: "=") else { continue }
            let name = field[..<eq].trimmingCharacters(in: .whitespaces).lowercased()
            var v = field[field.index(after: eq)...].trimmingCharacters(in: .whitespaces)
            if v.hasPrefix("\"") && v.hasSuffix("\"") && v.count >= 2 { v = String(v.dropFirst().dropLast()) }
            out.params[name] = v
        }
        return out
    }

    static func encoding(_ header: String) -> String {
        value(of: "content-transfer-encoding", in: header)?
            .trimmingCharacters(in: .whitespaces).lowercased() ?? ""
    }

    private static func value(of name: String, in header: String) -> String? {
        for line in MessageHeaders.unfolded(header) {
            guard let colon = line.firstIndex(of: ":") else { continue }
            if line[..<colon].lowercased().trimmingCharacters(in: .whitespaces) == name {
                return String(line[line.index(after: colon)...])
                    .trimmingCharacters(in: .whitespaces)
            }
        }
        return nil
    }

    /// Split on the semicolons that separate parameters, ignoring the
    /// ones inside a quoted value — a boundary may contain one, and
    /// splitting there loses the rest of it.
    private static func splitOnSemicolons(_ s: String) -> [String] {
        var out: [String] = []
        var current = ""
        var quoted = false
        for ch in s {
            if ch == "\"" { quoted.toggle() }
            if ch == ";" && !quoted {
                out.append(current.trimmingCharacters(in: .whitespaces))
                current = ""
                continue
            }
            current.append(ch)
        }
        out.append(current.trimmingCharacters(in: .whitespaces))
        return out.filter { !$0.isEmpty }
    }

    // MARK: - bytes

    /// Header block and body, at the first blank line.
    static func split(_ raw: Data) -> (Data, Data) {
        if let r = ranges(of: Data("\r\n\r\n".utf8), in: raw).first {
            return (raw[raw.startIndex..<r], raw[(r + 4)...])
        }
        if let r = ranges(of: Data("\n\n".utf8), in: raw).first {
            return (raw[raw.startIndex..<r], raw[(r + 2)...])
        }
        return (raw, Data())
    }

    private static func skipLine(_ d: Data, _ from: Int) -> Int {
        var i = from
        while i < d.endIndex && d[i] != 0x0A { i += 1 }
        return min(i + 1, d.endIndex)
    }

    private static func ranges(of needle: Data, in haystack: Data) -> [Int] {
        guard !needle.isEmpty, haystack.count >= needle.count else { return [] }
        var out: [Int] = []
        let last = haystack.endIndex - needle.count
        var i = haystack.startIndex
        while i <= last {
            if haystack[i] == needle[needle.startIndex] {
                var match = true
                for k in 0..<needle.count where haystack[i + k] != needle[needle.startIndex + k] {
                    match = false
                    break
                }
                if match { out.append(i) }
            }
            i += 1
        }
        return out
    }
}

File: android/app/src/main/kotlin/jp/golia/mailrs/accounts/MessageBody.kt
package jp.golia.mailrs.accounts

import java.nio.charset.Charset
import java.util.Base64

/**
 * The readable part of a message, out of its raw bytes.
 *
 * Not a full MIME implementation and not trying to be: what a reader
 * needs is the one part worth showing and the text of it. Attachments,
 * signatures and the rest of the tree are left alone.
 *
 * Bytes rather than a `String` throughout, and that is the whole reason
 * this exists separately from [MessageHeaders]. A message says what its
 * charset is *inside itself*; decoding it as UTF-8 on the way in — which
 * is what reading a socket into a `String` does — turns every Shift_JIS
 * and windows-1252 message into replacement characters before anything
 * has read the header that would have said so.
 */
object MessageBody {
    data class Display(val text: String, val isHtml: Boolean) {
        companion object {
            val EMPTY = Display("", false)
        }
    }

    /** The part worth showing. */
    fun extract(raw: ByteArray): Display {
        val (header, body) = split(raw)
        return part(String(header, Charsets.UTF_8), body)
    }

    private fun part(header: String, body: ByteArray): Display {
        val type = contentType(header)
        if (type.type == "multipart") {
            val boundary = type.params["boundary"]
            if (boundary.isNullOrEmpty()) {
                // A multipart with no boundary cannot be taken apart.
                // Showing the raw source beats showing nothing: the text
                // is usually in there, sandwiched between header blocks
                // a person can read past. Deliberately not via
                // `decoded`, which would throw it away for not being
                // `text/*` — it is not text/* by declaration only.
                return Display(
                    string(decodeTransfer(body, encoding(header)), type.params["charset"]),
                    false,
                )
            }
            return choose(pieces(body, boundary), type.subtype)
        }
        return decoded(header, body)
    }

    /**
     * Which of a multipart's pieces to show.
     *
     * `alternative` is the same message written twice, so the choice is
     * a preference: plain text first, markup only when there is no plain
     * text. Every other kind — `mixed`, `related`, `signed` — is a
     * message plus its attachments, and the first piece with anything
     * readable in it is the message.
     */
    private fun choose(parts: List<ByteArray>, kind: String): Display {
        val shown = parts.map { piece ->
            val (h, b) = split(piece)
            part(String(h, Charsets.UTF_8), b)
        }
        if (kind == "alternative") {
            shown.firstOrNull { !it.isHtml && it.text.isNotEmpty() }?.let { return it }
        }
        return shown.firstOrNull { it.text.isNotEmpty() } ?: Display.EMPTY
    }

    /**
     * The pieces between the boundary delimiters.
     *
     * Everything before the first delimiter is the preamble and
     * everything after the closing one is the epilogue; both are there
     * for mail readers that cannot do MIME at all, and neither is part
     * of the message.
     */
    private fun pieces(body: ByteArray, boundary: String): List<ByteArray> {
        val delimiter = "--$boundary".toByteArray()
        // The delimiter must begin a line, or a boundary string that
        // happens to appear inside a part cuts it in half.
        val starts = indexesOf(delimiter, body).filter { it == 0 || body[it - 1] == '\n'.code.toByte() }
        if (starts.isEmpty()) return emptyList()
        val out = mutableListOf<ByteArray>()
        for ((i, start) in starts.withIndex()) {
            val after = start + delimiter.size
            // The closing delimiter: nothing after it is ours.
            if (after + 1 < body.size && body[after] == '-'.code.toByte() &&
                body[after + 1] == '-'.code.toByte()
            ) {
                break
            }
            if (i + 1 >= starts.size) break
            val from = skipLine(body, after)
            val to = starts[i + 1]
            if (from < to) out.add(body.copyOfRange(from, to))
        }
        return out
    }

    private fun decoded(header: String, body: ByteArray): Display {
        val type = contentType(header)
        // Anything that is not text is not something to show as text. An
        // attached PDF decoded as if it were a message reads as a screen
        // of noise.
        if (type.type.isNotEmpty() && type.type != "text") return Display.EMPTY
        val bytes = decodeTransfer(body, encoding(header))
        return Display(string(bytes, type.params["charset"]), type.subtype == "html")
    }

    private fun decodeTransfer(body: ByteArray, cte: String): ByteArray = when (cte) {
        "base64" -> Base64Body.decode(body)
        "quoted-printable" -> QuotedPrintable.decode(body)
        else -> body
    }

    /**
     * Bytes to text, in the charset the message declared.
     *
     * UTF-8 when nothing was declared — it is what most mail is, and it
     * is self-checking, so a wrong guess fails loudly enough to fall
     * back rather than producing plausible nonsense.
     */
    private fun string(bytes: ByteArray, charset: String?): String {
        if (charset.isNullOrEmpty()) return String(bytes, Charsets.UTF_8)
        return try {
            String(bytes, Charset.forName(charset))
        } catch (_: Exception) {
            String(bytes, Charsets.UTF_8)
        }
    }

    data class ContentType(
        val type: String = "",
        val subtype: String = "",
        val params: Map<String, String> = emptyMap(),
    )

    fun contentType(header: String): ContentType {
        val raw = valueOf("content-type", header) ?: return ContentType("text", "plain")
        val fields = splitOnSemicolons(raw)
        val full = fields.firstOrNull().orEmpty()
        val halves = full.split("/", limit = 2)
        val params = mutableMapOf<String, String>()
        for (field in fields.drop(1)) {
            val eq = field.indexOf('=')
            if (eq < 0) continue
            val name = field.substring(0, eq).trim().lowercase()
            var v = field.substring(eq + 1).trim()
            if (v.length >= 2 && v.startsWith("\"") && v.endsWith("\"")) v = v.substring(1, v.length - 1)
            params[name] = v
        }
        return ContentType(
            halves.getOrNull(0)?.trim()?.lowercase().orEmpty(),
            halves.getOrNull(1)?.trim()?.lowercase().orEmpty(),
            params,
        )
    }

    private fun encoding(header: String): String =
        valueOf("content-transfer-encoding", header)?.trim()?.lowercase().orEmpty()

    private fun valueOf(name: String, header: String): String? {
        for (line in MessageHeaders.unfolded(header)) {
            val colon = line.indexOf(':')
            if (colon < 0) continue
            if (line.substring(0, colon).trim().lowercase() == name) {
                return line.substring(colon + 1).trim()
            }
        }
        return null
    }

    /**
     * Split on the semicolons that separate parameters, ignoring the
     * ones inside a quoted value — a boundary may contain one, and
     * splitting there loses the rest of it.
     */
    private fun splitOnSemicolons(s: String): List<String> {
        val out = mutableListOf<String>()
        val current = StringBuilder()
        var quoted = false
        for (ch in s) {
            if (ch == '"') quoted = !quoted
            if (ch == ';' && !quoted) {
                out.add(current.toString().trim())
                current.clear()
                continue
            }
            current.append(ch)
        }
        out.add(current.toString().trim())
        return out.filter { it.isNotEmpty() }
    }

    /** Header block and body, at the first blank line. */
    fun split(raw: ByteArray): Pair<ByteArray, ByteArray> {
        indexesOf("\r\n\r\n".toByteArray(), raw).firstOrNull()?.let {
            return raw.copyOfRange(0, it) to raw.copyOfRange(it + 4, raw.size)
        }
        indexesOf("\n\n".toByteArray(), raw).firstOrNull()?.let {
            return raw.copyOfRange(0, it) to raw.copyOfRange(it + 2, raw.size)
        }
        return raw to ByteArray(0)
    }

    private fun skipLine(d: ByteArray, from: Int): Int {
        var i = from
        while (i < d.size && d[i] != '\n'.code.toByte()) i++
        return minOf(i + 1, d.size)
    }

    private fun indexesOf(needle: ByteArray, haystack: ByteArray): List<Int> {
        if (needle.isEmpty() || haystack.size < needle.size) return emptyList()
        val out = mutableListOf<Int>()
        for (i in 0..haystack.size - needle.size) {
            var match = true
            for (k in needle.indices) {
                if (haystack[i + k] != needle[k]) {
                    match = false
                    break
                }
            }
            if (match) out.add(i)
        }
        return out
    }
}

/** RFC 2045 §6.7. */
object QuotedPrintable {
    fun decode(input: ByteArray): ByteArray {
        val out = java.io.ByteArrayOutputStream(input.size)
        var i = 0
        while (i < input.size) {
            val byte = input[i]
            if (byte != '='.code.toByte()) {
                out.write(byte.toInt())
                i++
                continue
            }
            // A soft line break: `=` at end of line means the line was
            // wrapped there and there is no character at all.
            if (i + 2 < input.size && input[i + 1] == '\r'.code.toByte() &&
                input[i + 2] == '\n'.code.toByte()
            ) {
                i += 3
                continue
            }
            if (i + 1 < input.size && input[i + 1] == '\n'.code.toByte()) {
                i += 2
                continue
            }
            val hi = if (i + 2 < input.size) hex(input[i + 1]) else null
            val lo = if (i + 2 < input.size) hex(input[i + 2]) else null
            if (hi != null && lo != null) {
                out.write(hi shl 4 or lo)
                i += 3
                continue
            }
            // A lone `=` is not valid, and the choice matters: dropping
            // it silently loses a character somebody typed, so it is
            // kept as itself.
            out.write(byte.toInt())
            i++
        }
        return out.toByteArray()
    }

    private fun hex(b: Byte): Int? = when (val c = b.toInt().toChar()) {
        in '0'..'9' -> c - '0'
        in 'A'..'F' -> c - 'A' + 10
        in 'a'..'f' -> c - 'a' + 10
        else -> null
    }
}

/**
 * Base64 as it arrives in mail: wrapped across lines, and sometimes with
 * characters no encoder should have emitted.
 */
object Base64Body {
    fun decode(input: ByteArray): ByteArray {
        val cleaned = input.filter { b ->
            val c = b.toInt().toChar()
            c in 'A'..'Z' || c in 'a'..'z' || c in '0'..'9' || c == '+' || c == '/' || c == '='
        }.toByteArray()
        var text = String(cleaned, Charsets.US_ASCII).trimEnd('=')
        // Padding is often missing from the last line. Without this a
        // whole message decodes to nothing rather than to itself.
        when (text.length % 4) {
            2 -> text += "=="
            3 -> text += "="
            1 -> text = text.dropLast(1)
        }
        return try {
            Base64.getDecoder().decode(text)
        } catch (_: IllegalArgumentException) {
            ByteArray(0)
        }
    }
}
